src/plugins/hetero/src/compiled_model.cpp

```diff
@@ -0,0 +1,762 @@
+                                                      intermed_input);
+                        if (in_it != orderedSubgraphs[id2]._parameters.end()) {
+                            auto out_idx = std::distance(orderedSubgraphs[id]._results.begin(), out_it);
+                            auto in_idx = std::distance(orderedSubgraphs[id2]._parameters.begin(), in_it);
+                            m_submodels_output_to_input[{id, out_idx}] = {id2, in_idx};
+                        }
+                    }
+                }
+            }
+        }
+
+        m_networks.resize(orderedSubgraphs.size());
+        std::vector<std::shared_ptr<ov::Model>> subFunctions(orderedSubgraphs.size());
+        int id = 0;
+        for (auto&& subgraph : orderedSubgraphs) {
+            m_networks[id]._device = subgraph._affinity;
+            subFunctions[id] = std::make_shared<ov::Model>(subgraph._results,
+                                                           subgraph._sinks,
+                                                           subgraph._parameters,
+                                                           m_name + '_' + std::to_string(id));
+            m_networks[id]._clonedNetwork = subFunctions[id]->clone();  // TODO vurusovs IS CLONE REQUIRED?
+            ++id;
+        }
+        for (auto&& network : m_networks) {
+            auto metaDevices =
+                get_hetero_plugin()->get_properties_per_device(network._device, m_cfg.GetDeviceProperties());
+
+            // disable caching for subgraphs, because the whole HETERO model is cached
+            auto device_config = metaDevices[network._device];
+            device_config[ov::cache_dir.name()] = "";
+
+            network._network =
+                plugin->get_core()->compile_model(network._clonedNetwork, network._device, device_config);
+        }
+
+        // Restore inputs/outputs from compiled models
+        m_compiled_inputs.reserve(m_inputs_to_submodel_inputs.size());
+        for (const auto& it : m_inputs_to_submodel_inputs) {
+            const auto& submodel_idx = it.first;
+            const auto& input_idx = it.second;
+            m_compiled_inputs.emplace_back(m_networks[submodel_idx]._network->inputs()[input_idx]);
+        }
+        m_compiled_outputs.reserve(m_outputs_to_submodel_outputs.size());
+        for (const auto& it : m_outputs_to_submodel_outputs) {
+            const auto& submodel_idx = it.first;
+            const auto& output_idx = it.second;
+            m_compiled_outputs.emplace_back(m_networks[submodel_idx]._network->outputs()[output_idx]);
+        }
+
+    } catch (const InferenceEngine::Exception& e) {
+        // Some transformations can throw legacy exception
+        OPENVINO_THROW(e.what());
+    } catch (const std::exception& e) {
+        OPENVINO_THROW("Standard exception from compilation library: ", e.what());
+    } catch (...) {
+        OPENVINO_THROW("Generic exception is thrown");
+    }
```

Comment:
I think we need to provide more detailed messages.
Also rethrow ov::Exception.

And can we avoid `catch (...)`?

Reply:
https://github.com/openvinotoolkit/openvino/pull/18429/files#r1259447405